### PR TITLE
[SYCL] Add CMake macro for building targets with device code

### PR DESCRIPTION
**Pull request by alexbatashev**
The `add_sycl_executable` can be used to create device code-enabled unit tests.

Signed-off-by: Alexander Batashev <alexander.batashev@intel.com>

### CODE REVIEW

**Comment by alexbatashev** [comment]
@intel/llvm-reviewers-runtime ping.

File: sycl/cmake/modules/AddSYCLDeviceTarget.cmake [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      set(CXX_COMPILER clang-cl.exe)
+  endif()
+  set(DEVICE_COMPILER_EXECUTABLE ${LLVM_RUNTIME_OUTPUT_INTDIR}/${CXX_COMPILER})
+
+  add_custom_target(${ARG_TARGET_NAME}_exec ALL
+    COMMAND ${DEVICE_COMPILER_EXECUTABLE} -fsycl ${ARG_OPTIONS} ${ARG_SOURCES}
+      -o ${CMAKE_CURRENT_BINARY_DIR}/${ARG_TARGET_NAME}
+    BYPRODUCTS ${CMAKE_CURRENT_BINARY_DIR}/${ARG_TARGET_NAME}
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+  add_dependencies(${ARG_TARGET_NAME}_exec sycl clang)
```
Comment (s-kanaev) [inline sycl/cmake/modules/AddSYCLDeviceTarget.cmake:19]: Why do you need a dependency on clang and not clang++?
Also, if you use clang you may employ driver-mode=g++ command line argument.

Comment (alexbatashev) [reply sycl/cmake/modules/AddSYCLDeviceTarget.cmake:19]: `clang++` is only a symlink to `clang`. I believe this should be fine.

**Comment by alexbatashev** [comment]
> Looks good.
> Should there be a macro to add library?

I'm not sure. Personally, I don't need SYCL-enabled libraries at the moment. But if there's such a need, I can enhance this patch further.

File: sycl/cmake/modules/AddSYCLDeviceTarget.cmake (hunk 1 above)
Comment (bader) [reply sycl/cmake/modules/AddSYCLDeviceTarget.cmake:19]: > `clang++` is only a symlink to `clang`. I believe this should be fine.

What about Windows?

Comment (alexbatashev) [reply sycl/cmake/modules/AddSYCLDeviceTarget.cmake:19]: Fair point. I'll double check this on windows.

Comment (bader) [reply sycl/cmake/modules/AddSYCLDeviceTarget.cmake:19]: I think you are fine. If I understand this line correctly, here we add dependency on cmake target, not executable name and AFAIK, clang cmake target builds all drivers: clang, clang++ and clang-cl.

**Comment by bader** [comment]
@alexbatashev, are you going to modify this patch to address @s-kanaev comments are it's ready for merge?

**Comment by alexbatashev** [comment]
> @alexbatashev, are you going to modify this patch to address @s-kanaev comments are it's ready for merge?

No, I'm not going to add support for device code for libraries. At the moment I don't see any use for them and I don't want to add dead code. As mentioned in PR description, this patch has a specific purpose to enable SYCL device code in unit tests.

File: sycl/cmake/modules/AddSYCLDeviceTarget.cmake [hunk 2]
```diff
@@ -0,0 +1,24 @@
+macro(add_sycl_executable ARG_TARGET_NAME)
```
Comment (bader) [inline sycl/cmake/modules/AddSYCLDeviceTarget.cmake:1]: I would name it `AddSYCLExecutable.cmake` to match the macro defined in this file.

File: sycl/cmake/modules/AddSYCLDeviceTarget.cmake (hunk 1 above)
Comment (bader) [reply sycl/cmake/modules/AddSYCLDeviceTarget.cmake:19]: @alexbatashev, @s-kanaev, is this resolved?

File: sycl/cmake/modules/AddSYCLDeviceTarget.cmake (hunk 2 above)
Comment (alexbatashev) [reply sycl/cmake/modules/AddSYCLDeviceTarget.cmake]: Done

Comment (bader) [reply sycl/cmake/modules/AddSYCLDeviceTarget.cmake]: I think you forgot to update the name in sycl/CMakeLists.txt

**Pull request merged**